Index: api/baseapi.cpp
===================================================================
--- api/baseapi.cpp	(revision 866)
+++ api/baseapi.cpp	(working copy)
@@ -200,11 +200,16 @@
  * be returned.
  * @return: 0 on success and -1 on initialization failure.
  */
+
+#include <locale.h> 
+
 int TessBaseAPI::Init(const char* datapath, const char* language,
                       OcrEngineMode oem, char **configs, int configs_size,
                       const GenericVector<STRING> *vars_vec,
                       const GenericVector<STRING> *vars_values,
                       bool set_only_non_debug_params) {
+  setlocale(LC_NUMERIC, "C");
+
   // Default language is "eng".
   if (language == NULL) language = "eng";
   // If the datapath, OcrEngineMode or the language have changed - start again.
@@ -245,6 +250,7 @@
   // For same language and datapath, just reset the adaptive classifier.
   if (reset_classifier) tesseract_->ResetAdaptiveClassifier();
 
+
   return 0;
 }
 
